modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/queue/KafkaConsumer.scala

```diff
@@ -1,56 +1,134 @@
-            index(event).map(_ => msg.committableOffset)
-          case Left(e) =>
-            log.error(s"Failed to decode message: $msg", e)
-            Future.failed(e)
+        val process = self ? msg.record.value
+        process.map {
+          case Done => msg.committableOffset
+          case _ => throw new IllegalStateException(s"Unexpected response after processing message: $msg")
         }
       }
       .batch(max = 20, first => CommittableOffsetBatch.empty.updated(first)) { (batch, elem) =>
         batch.updated(elem)
       }
       .mapAsync(1)(_.commitScaladsl())
       .runWith(Sink.ignore)
+
+    // $COVERAGE-OFF$
+    done.onComplete {
+      case Failure(e) =>
+        log.error("Stream failed with an error, stopping the actor and restarting.", e)
+        self ! PoisonPill
+      case Success(_) =>
+        log.warn("Kafka consumer stream ended gracefully, however this should not happen; restarting.")
+        self ! PoisonPill
+    }
+    // $COVERAGE-ON$
+  }
+
+}
+
+object KafkaConsumer {
+
+  /**
+    * Starts a Kafka consumer that reads messages from a particular ''topic'', decodes them into
+    * events of type ''T'' and indexes them using the provided function.  The consumer stream is
+    * wrapped in its own actor whose life-cycle is managed by a [[akka.pattern.BackoffSupervisor]] instance.
+    *
+    * @param settings an instance of [[akka.kafka.ConsumerSettings]]
+    * @param index    the indexing function that is applied to received events
+    * @param topic    the Kafka topic to read messages from
+    * @param decoder  a Circe decoder to deserialize received messages
+    * @param as       an implicitly available actor system
+    * @tparam T       the event type
+    * @return the supervisor actor handle
+    * @note  Calling this method multiple times within the same actor system context will result in
+    *        an [[akka.actor.InvalidActorNameException]] being thrown. You must stop any previously existing
+    *        supervisor first.
+    */
+  def start[T](settings: ConsumerSettings[String, String],
+               index: T => Future[Unit],
+               topic: String,
+               decoder: Decoder[T])(implicit as: ActorSystem): ActorRef = {
+    val childProps = Props(classOf[KafkaConsumerWrapper[T]], settings, index, topic, decoder)
+    val supervisor = as.actorOf(
+      BackoffSupervisor.props(
+        Backoff.onStop(
+          childProps,
+          childName = "kafka-stream-supervised-actor",
+          minBackoff = 3.seconds,
+          maxBackoff = 30.seconds,
+          randomFactor = 0.2
+        )),
+      name = "kafka-stream-supervisor"
+    )
+
+    supervisor
+  }
+
+  /**
+    * Kills the managed actor that wraps the Kafka consumer, which should be re-spawned by its supervisor.
+    *
+    * @param supervisor the supervisor actor handle
+    */
+  @VisibleForTesting
+  private[queue] def stop(supervisor: ActorRef)(implicit as: ActorSystem): Unit = {
```

Comment:
This method should be on the tests if it is used only when testing, right?
Besides that, `scapegoat` is complaining because of the use of `ref.get`